src/tools/storage-key-recipe-resolver.ts

```diff
@@ -0,0 +1,127 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Runtime} from '../runtime/runtime.js';
+import {Manifest} from '../runtime/manifest.js';
+import {Loader} from '../platform/loader-web.js';
+import {IsValidOptions, Recipe, RecipeComponent} from '../runtime/recipe/recipe.js';
+import {ramDiskStorageKeyPrefixForTest} from '../runtime/testing/handle-for-test.js';
+import {Arc} from '../runtime/arc.js';
+import {RecipeResolver} from '../runtime/recipe/recipe-resolver.js';
+import {CapabilitiesResolver} from '../runtime/capabilities-resolver.js';
+import {Store} from '../runtime/storageNG/store.js';
+import {Exists} from '../runtime/storageNG/drivers/driver.js';
+
+/**
+ * Responsible for resolving recipes with storage keys.
+ */
+export class StorageKeyRecipeResolver {
+  private readonly runtime: Runtime;
+
+  constructor(context: Manifest) {
+    const loader = new Loader();
+    this.runtime = new Runtime({loader, context});
+  }
+
+  /**
+   * Produces resolved recipes with storage keys.
+   *
+   * TODO(alxr): Apply to long-running recipes appropriately.
+   * @throws Error if recipe fails to resolve on first or second pass.
+   * @yields Resolved recipes with storage keys
+   */
+  async* resolve(): AsyncGenerator<Recipe> {
+    for (const recipe of this.runtime.context.allRecipes) {
+      const arc = this.runtime.newArc(this.getArcId(recipe), ramDiskStorageKeyPrefixForTest());
+      const opts = {errors: new Map<Recipe | RecipeComponent, string>()};
+      const resolved = await this.resolveOrNormalize(recipe, arc, opts);
+      if (!resolved) {
+        throw Error(`Recipe ${recipe.name} failed to resolve:\n${[...opts.errors.values()].join('\n')}`);
+      }
+      this.createStoresForCreateHandles(resolved, arc);
+      resolved.normalize();
+      if (!resolved.isResolved()) {
+        throw Error(`Recipe ${resolved.name} did not properly resolve!\n${resolved.toString({showUnresolved: true})}`);
+      }
+      yield resolved;
+    }
+  }
+
+  /**
+   * Resolves unresolved recipe or normalizes resolved recipe.
+   *
+   * @param recipe long-running or ephemeral recipe
+   * @param arc Arc is associated with input recipe
+   * @param opts contains `errors` map for reporting.
+   */
+  async resolveOrNormalize(recipe: Recipe, arc: Arc, opts?: IsValidOptions): Promise<Recipe | null> {
```

Comment:
Recipe has to be normalized to be resolved. See `recipe.isResolved()`, the first thing it checks is whether a recipe is normalized.

What your method is doing is really:
1) Check whether a recipe is resolved.
2) If not, try resolving it with a recipe resolver.

The normalization is a prerequisite for resolution. Therefore I would change the name to: `tryResolve()`

Reply:
Thanks! I had trouble thinking of a good name.